File: frameworks/go/fiber/Implementation/middleware.go
package implementation

import (
	"github.com/gofiber/fiber/v3"
)

// middlewareRoutes put no-op layers in front of the handler. Fiber runs a route's handlers in
// the order they are registered, and each reaches the next with c.Next, so the layers come
// first and are on these two routes alone.
func middlewareRoutes(app *fiber.App, p *Payloads) {
	small := func(c fiber.Ctx) error { return c.JSON(&p.Small) }
	four := layered(4, small)
	sixteen := layered(16, small)

	app.Get("/middleware/none", small)

	app.Get("/middleware/four", four[0], four[1:]...)

	app.Get("/middleware/sixteen", sixteen[0], sixteen[1:]...)
}

// rb:wiring middleware.*
// noop is one layer: it calls the next handler and does nothing else.
func noop(c fiber.Ctx) error { return c.Next() }

// layered is the handler behind count no-op layers, in the order a route runs them.
func layered(count int, handler fiber.Handler) []any {
	chain := make([]any, count, count+1)
	for i := range chain {
		chain[i] = noop
	}
	return append(chain, handler)
}

// rb:end

File: frameworks/node/koa/Implementation/routes/middleware.ts
import type { Middleware } from "koa";

import type { Routes } from "../app.ts";

// rb:wiring middleware.*
/**
 * Koa's middleware is an async function that awaits next(), and @koa/router takes middleware in
 * front of a route's handler, so these run on one route and no other.
 */
const noop: Middleware = async (_ctx, next) => {
  await next();
};

const layers = (count: number) => Array.from({ length: count }, () => noop);
// rb:end

/** middleware: no-op layers in front of the handler. */
const middleware: Routes = (router, { payloads: p }) => {
  router.get("/middleware/none", (ctx) => {
    ctx.body = p.small;
  });

  router.get("/middleware/four", ...layers(4), (ctx) => {
    ctx.body = p.small;
  });

  router.get("/middleware/sixteen", ...layers(16), (ctx) => {
    ctx.body = p.small;
  });
};

export default middleware;
